File: backend/app/services/cache.py
import hashlib
import json
import time
from typing import Any

from redis.asyncio import Redis

from app.config import settings

_redis: Redis | None = None
_memory_cache: dict[str, tuple[float, Any]] = {}
_memory_versions: dict[str, int] = {}
_memory_locks: dict[str, float] = {}


async def _get_redis() -> Redis | None:
    global _redis
    if not settings.use_redis or not settings.redis_url:
        return None
    if _redis is None:
        _redis = Redis.from_url(settings.redis_url, decode_responses=True)
    return _redis


def _stable_key(prefix: str, payload: dict[str, Any]) -> str:
    digest = hashlib.sha256(json.dumps(payload, sort_keys=True, default=str).encode("utf-8")).hexdigest()
    return f"{prefix}:{digest}"


async def get_user_cache_version(user_id: str) -> int:
    redis = await _get_redis()
    key = f"unisync:cache:version:{user_id}"
    if redis:
        value = await redis.get(key)
        return int(value) if value else 0
    return _memory_versions.get(key, 0)


async def bump_user_cache_version(user_id: str) -> int:
    redis = await _get_redis()
    key = f"unisync:cache:version:{user_id}"
    if redis:
        value = await redis.incr(key)
        await redis.expire(key, 86400)
        return int(value)
    _memory_versions[key] = _memory_versions.get(key, 0) + 1
    return _memory_versions[key]


async def get_cached_json(prefix: str, payload: dict[str, Any]) -> Any | None:
    redis = await _get_redis()
    key = _stable_key(prefix, payload)
    if redis:
        raw = await redis.get(key)
        return json.loads(raw) if raw else None
    cached = _memory_cache.get(key)
    if not cached:
        return None
    expires_at, value = cached
    if expires_at <= time.time():
        _memory_cache.pop(key, None)
        return None
    return value


async def set_cached_json(prefix: str, payload: dict[str, Any], value: Any, ttl_seconds: int = 20) -> None:
    redis = await _get_redis()
    key = _stable_key(prefix, payload)
    if redis:
        await redis.set(key, json.dumps(value), ex=ttl_seconds)
        return
    _memory_cache[key] = (time.time() + ttl_seconds, value)


async def acquire_recent_lock(scope: str, identity: str, ttl_seconds: int = 900) -> bool:
    redis = await _get_redis()
    key = f"unisync:lock:{scope}:{identity}"
    if redis:
        return bool(await redis.set(key, "1", ex=ttl_seconds, nx=True))
    now = time.time()
    expires_at = _memory_locks.get(key, 0)
    if expires_at > now:
        return False
    _memory_locks[key] = now + ttl_seconds
    return True

File: backend2/src/services/cache.ts
import { createHash } from "node:crypto";

import IORedis from "ioredis";

import { settings } from "../config.js";

type RedisClient = ReturnType<typeof createRedis>;
let redis: RedisClient | null = null;
const memoryCache = new Map<string, { expiresAt: number; value: unknown }>();
const memoryVersions = new Map<string, number>();
const memoryLocks = new Map<string, number>();

function stableKey(prefix: string, payload: Record<string, unknown>): string {
  const digest = createHash("sha256").update(JSON.stringify(payload)).digest("hex");
  return `${prefix}:${digest}`;
}

function createRedis(url: string) {
  const RedisCtor = IORedis as unknown as new (redisUrl: string) => any
  return new RedisCtor(url)
}

function getRedis(): RedisClient | null {
  if (!settings.useRedis || !settings.redisUrl) return null;
  if (!redis) redis = createRedis(settings.redisUrl);
  return redis;
}

export async function getUserCacheVersion(userId: string): Promise<number> {
  const cache = getRedis();
  const key = `unisync:cache:version:${userId}`;
  if (cache) {
    const value = await cache.get(key);
    return value ? Number(value) : 0;
  }
  return memoryVersions.get(key) ?? 0;
}

export async function bumpUserCacheVersion(userId: string): Promise<number> {
  const cache = getRedis();
  const key = `unisync:cache:version:${userId}`;
  if (cache) {
    const value = await cache.incr(key);
    await cache.expire(key, 86400);
    return value;
  }
  const next = (memoryVersions.get(key) ?? 0) + 1;
  memoryVersions.set(key, next);
  return next;
}

export async function getCachedJson<T>(prefix: string, payload: Record<string, unknown>): Promise<T | null> {
  const cache = getRedis();
  const key = stableKey(prefix, payload);
  if (cache) {
    const raw = await cache.get(key);
    return raw ? (JSON.parse(raw) as T) : null;
  }
  const entry = memoryCache.get(key);
  if (!entry || entry.expiresAt <= Date.now()) {
    memoryCache.delete(key);
    return null;
  }
  return entry.value as T;
}

export async function setCachedJson(
  prefix: string,
  payload: Record<string, unknown>,
  value: unknown,
  ttlSeconds = 20
): Promise<void> {
  const cache = getRedis();
  const key = stableKey(prefix, payload);
  if (cache) {
    await cache.set(key, JSON.stringify(value), "EX", ttlSeconds);
    return;
  }
  memoryCache.set(key, { expiresAt: Date.now() + ttlSeconds * 1000, value });
}

export async function acquireRecentLock(scope: string, identity: string, ttlSeconds = 900): Promise<boolean> {
  const cache = getRedis();
  const key = `unisync:lock:${scope}:${identity}`;
  if (cache) {
    const result = await cache.set(key, "1", "EX", ttlSeconds, "NX");
    return result === "OK";
  }
  const now = Date.now();
  const expiresAt = memoryLocks.get(key) ?? 0;
  if (expiresAt > now) return false;
  memoryLocks.set(key, now + ttlSeconds * 1000);
  return true;
}
